### PR TITLE
fix(ngAnimate): correctly animate transcluded clones with `templateUrl`

### LINKED ISSUE #15510: Enter/leave animation not working for for first item in ng-repeat
If an element has both ng-repeat and a directive that uses templateUrl `<div my-dir ng-repeat="...">`, the enter animation doesn't work for the first item, even if the template is in $templateCache.

[Demo](http://plnkr.co/edit/j86chAryADuQsSM8ctte?p=preview)

If you don't pre-warm the $templateCache, the leave animation doesn't work either. You can comment out the run block in the demo to see this.

If the directive is on a child element, the problem goes away, and all animations work. And if fetching the template is fast enough, enter and leave animations both work from a cold cache.

    <div ng-repeat="x in y"> <div my-dir></div> </div>

I tested this in Angular 1.5.8 and 1.6.0, with Chrome 54, Safari 10, and Firefox 50.

@gkalpak suggested I make this ticket in #12612.

### PR DESCRIPTION
**Pull request by gkalpak**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix.

**What is the current behavior? (You can also link to an open issue here)**
`$animate` decides whether an animation should be cancelled based on some assumption that don't hold in specific cases (e.g. when animating transcluded clones with `templateUrl` directives on them for the first time). As a result, the entering elements will not be animated in such cases. This affects commonly used, structural built-in directives (`ngIf`, `ngRepeat`, `ngSwitch` etc).
See also #15510.

**What is the new behavior (if this is a feature change)?**
All entering animations work as expected (from the first time).

**Does this PR introduce a breaking change?**
No.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

**Other information**:
BTW, I believe this is a side effect of lazily compiling transcluded content.

In addition to the last `fix` commit that fixes the bug, this PR also includes several `refactor` commits that simplify `$animate` internally, mainly by removing unnecessary wrapping/unwrapping of DOM nodes in `jqLite`/`jQuery` collections. This changes may also positively affect performance, especially in animation-heavy apps (~~although I haven't benchmarked it - yet~~ it does indeed :wink:).
(It is probably easier to review one commit at a time.)

Party addresses #14074 and #14124.

Fixes #15510.

### CODE REVIEW

**Comment by gkalpak** [comment]
I added one more commit with a (very basic) animation benchmark. It simply `$animate.enter`s and immediatelly `$animate.leave`s 1000 elements - via `ngRepeat`. With the changes proposed in this commit, the benchmark shows a 5%-7% reduction in total execution time (which is not bad at all for a bunch of refatoring commits 😛 - I am tempted to change them to `perf`).

<sub>(Take the results with a gran of salt, as the benchmark scenario is a little artificial.)</sub>

File: test/ngAnimate/animateSpec.js [hunk 1]
```diff
@@ -459,6 +459,26 @@ describe('animations', function() {
...
+      module(function($compileProvider) {
+        $compileProvider.directive('foo', valueFn({templateUrl: 'foo.html'}));
+      });
+
+      inject(function($animate, $compile, $rootScope, $templateCache) {
+        parent.append(jqLite('<foo ng-if="showFoo"></foo>'));
+        $templateCache.put('foo.html', '<div>FOO</div>');
+        $compile(parent)($rootScope);
+
+        expect(capturedAnimation).toBe(null);
```
Comment (mgol) [inline test/ngAnimate/animateSpec.js]: Everywhere in other places it's only checked if `capturedAnimation` is falsy. We should be consistent.

File: benchmarks/animation-bp/app.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+angular.
+  module('animationBenchmark', ['ngAnimate'], config).
```
Comment (mgol) [inline benchmarks/animation-bp/app.js]: I dislike this dot placement a lot. :P We use the leading dot in most places and I'd like to ESLint-enforce it one day...

File: benchmarks/animation-bp/main.html [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<style>
+  [ng-cloak] { display: none !important; }
+  .animation-container .ng-enter,
+  .animation-container .ng-leave {
+    transition: all 0.1s
```
Comment (mgol) [inline benchmarks/animation-bp/main.html]: A missing semi-colon;

**Review by mgol — CHANGES_REQUESTED** [review]

File: benchmarks/animation-bp/main.html (hunk 3 above)
Comment (gkalpak) [inline benchmarks/animation-bp/main.html]: 😱

File: src/ngAnimate/animateQueue.js [hunk 4]
```diff
@@ -610,65 +610,61 @@ var $$AnimateQueueProvider = ['$animateProvider', /** @this */ function($animate
...
-    function areAnimationsAllowed(element, parentElement, event) {
-      var bodyElement = jqLite($document[0].body);
-      var bodyElementDetected = isMatchingElement(element, bodyElement) || element[0].nodeName === 'HTML';
-      var rootElementDetected = isMatchingElement(element, $rootElement);
+    function areAnimationsAllowed(node, parentNode, event) {
+      var bodyNode = $document[0].body;
+      var rootNode = getDomNode($rootElement);
+
+      var bodyNodeDetected = (node === bodyNode) || node.nodeName === 'HTML';
+      var rootNodeDetected = (node === rootNode);
```
Comment (gkalpak) [inline src/ngAnimate/animateQueue.js]: A couple of reasons (at least):

1. I find them more readable. I think it is a remnant from my early Java days (and [this ancient document][1]).
2. It is also visually closer to `isMatchingElement(...)` 😛

<sub>(Hey, I changed the trailing dots, I am not removing the parens.)</sub>

[1]: http://www.oracle.com/technetwork/java/codeconventions-150003.pdf

File: test/ngAnimate/animateSpec.js (hunk 1 above)
Comment (gkalpak) [inline test/ngAnimate/animateSpec.js]: So, you want me to change all other places to `toBe(null)`? Fair enough...

File: benchmarks/animation-bp/app.js (hunk 2 above)
Comment (mgol) [inline benchmarks/animation-bp/app.js]: > Come on! I speeded up your animations 5%-7%. And all you care about is where I put a tiny little dot?

No, I like this PR a lot! I'm just providing feedback for some small issues. :)

Good job overall, though, perhaps I should have spelled that out. 👏

**Comment by gkalpak** [comment]
Addressed (most of) @mgol's comments.

File: test/ngAnimate/animateSpec.js (hunk 1 above)
Comment (mgol) [inline test/ngAnimate/animateSpec.js]: I was thinking the other way round but maybe your idea has more sense. ;)

In some cases it's initially `undefined`, though, not `null`, this file is not very consistent in that regard. Perhaps that's material for a separate PR (or at least a commit) but I just want us to be consistent. :)

File: test/ngAnimate/animateSpec.js [hunk 5]
```diff
@@ -168,7 +168,7 @@ describe('animations', function() {
       inject(function($animate, $rootScope) {
         $animate.enter(element, parent);
         $rootScope.$digest();
-        expect(capturedAnimation).toBeFalsy();
+        expect(capturedAnimation).toBe(null);
```
Comment (mgol) [inline test/ngAnimate/animateSpec.js]: I'd put those changes in a separate commit as they don't really have a lot in common with the original one.

File: test/ngAnimate/animateSpec.js (hunk 1 above)
Comment (mgol) [inline test/ngAnimate/animateSpec.js]: Ah, it's `undefined` only in one test:
https://github.com/gkalpak/angular.js/blob/acd9e7b757cdb9f04ae7a12ff21fdc8d27f314b5/test/ngAnimate/animateSpec.js#L21-L46

I guess the changes are fine then.

Comment (gkalpak) [inline test/ngAnimate/animateSpec.js]: I (blindly) changed all occurrences of `expect(capturedAnimation).toBeFalsy()` to `expect(capturedAnimation).toBe(null)` and the tests were passing :confused:

File: benchmarks/animation-bp/app.js (hunk 2 above)
Comment (gkalpak) [inline benchmarks/animation-bp/app.js]: <sub>Just in case it wasn't clear, I was just kidding (in an attempt to distract everyone from the fact that I was busted trying to sneak in those trailing dots 😛 </sub>

**Comment by gkalpak** [comment]
Rearranged the commits (and fixed a test failure).

File: src/ngAnimate/animateQueue.js (hunk 4 above)
Comment (Narretz) [inline src/ngAnimate/animateQueue.js]: Looks like we need an eslint rule for that :D But I prefer the parens too

Comment (mgol) [inline src/ngAnimate/animateQueue.js]: I like not having parents here but as long as ESLint will shout at me if I forget them I'd be satisfied as well. ;)

File: src/ngAnimate/animateQueue.js [hunk 6]
```diff
@@ -479,6 +479,15 @@ var $$AnimateQueueProvider = ['$animateProvider', /** @this */ function($animate
...
 
       $rootScope.$$postDigest(function() {
+        // It is possible that the DOM nodes inside `originalElement` have been replaced. This can
+        // happen if the animated element is a transcluded clone and also has a `templateUrl`
+        // directive on it. Therefore, we must recreate `element` in order to interact with the
+        // actual DOM nodes.
+        // Note: We still need to use the old `node` for certain things, such as looking up in
+        //       HashMaps where it was used as the key.
+
+        element = stripCommentsFromElement(originalElement);
```
Comment (Narretz) [inline src/ngAnimate/animateQueue.js]: I take it this has no performance implications? This now applies to all animations.

Comment (gkalpak) [inline src/ngAnimate/animateQueue.js]: I haven't benchmarked this change on its own. I assume it has an unnoticeable impact on performance (considering all the performance-heavy-ish stuff that is going on during animations). But overall, this PR offers a non-negligible performance gain (see https://github.com/angular/angular.js/pull/15514#issuecomment-267764426).

**Comment by gkalpak** [comment]
@mgol, @Narretz: PTAL

**Comment by Narretz** [comment]
LGTM
@mgol can you check this again?

**Review by mgol — APPROVED** [review]
